pennylane/resources/first_quantization.py

```diff
@@ -0,0 +1,221 @@
+
+
+def estimation_cost(lamb, error):
+    r"""Return the number of calls to the unitary needed to achieve the desired error in quantum
+    phase estimation.
+
+    Args:
+        lamb (float): 1-norm of a second-quantized Hamiltonian
+        error (float): target error in the algorithm
+
+    Returns:
+        int: number of calls to unitary
+
+    **Example**
+
+    >>> cost = estimation_cost(72.49779513025341, 0.001)
+    >>> print(cost)
+    113880
+    """
+    return int(np.ceil(np.pi * lamb / (2 * error)))
+
+
+def gate_cost(n, eta, omega, error, lamb, br=7, charge=0):
+    r"""Return the total number of Toffoli gates needed to implement the first quantization
+    algorithm.
+
+    Args:
+        n (int): number of basis states
+        eta (int): number of electrons
+        omega (float): unit cell volume
+        error (float): target error in the algorithm
+        lamb (float): 1-norm of the Hamiltonian
+        br (int): number of bits for ancilla qubit rotation
+        charge (int): total electric charge of the system
+
+    Returns:
+        int: the number of Toffoli gates needed to implement the first quantization algorithm
+
+    **Example**
+
+    >>> n = 100000
+    >>> eta = 156
+    >>> omega = 169.69608
+    >>> error = 0.01
+    >>> lamb = 5128920.595980267
+    >>> gate_cost(n, eta, omega, error, lamb)
+    10327614069516
+    """
+    e_cost = estimation_cost(lamb, error)
+    u_cost = unitary_cost(n, eta, omega, error, lamb, br, charge)
+
+    return e_cost * u_cost
+
+
+def qubit_cost(n, eta, omega, error, lamb, charge=0):
+    r"""Return the number of ancilla qubits needed to implement the first quantization algorithm.
+    ￼
+    The expression for computing the cost is taken from
+    [`arXiv:arXiv:2204.11890 <https://arxiv.org/abs/2204.11890>`_].
+
+    Args:
+        n (int): number of basis states
+        eta (int): number of electrons
+        omega (float): unit cell volume
+        error (float): target error in the algorithm
+        lamb (float): 1-norm of the Hamiltonian
+        charge (int): total electric charge of the system
+
+    Returns:
+        int: the number of ancilla qubits needed to implement the first quantization algorithm
+
+    **Example**
+
+    >>> n = 100000
+    >>> eta = 156
+    >>> omega = 169.69608
+    >>> error = 0.01
+    >>> lamb = 5128920.595980267
+    >>> qubit_cost(n, eta, omega, error, lamb)
+    4238
+    """
+    l_z = eta + charge
+    l_nu = 4 * np.pi * n ** (1 / 3)
+
+    n_p = np.ceil(np.log2(n ** (1 / 3) + 1))
+    n_t = np.log2(np.pi * lamb / (0.01 * error))
+    n_r = np.log2((eta * l_z * l_nu) / (0.01 * error * omega ** (1 / 3)))
+    n_m = np.log2(
+        (2 * eta) / (0.01 * error * omega ** (1 / 3))
+        + (eta - 1 + 2 * l_z)
+        + (7 * 2 ** (n_p + 1) - 9 * n_p - 11 - 3 * 2 ** (-1 * n_p))
+    )
+
+    qubits = 3 * eta * n_p + 4 * n_m * n_p + 12 * n_p
+    qubits += 2 * (np.ceil(np.log2(np.ceil(np.pi * lamb / (2 * error))))) + 5 * n_m
```

Comment:
we cannot use `0.01*error` for our finite precision arithmetic errors (call them `error_FPA`'s ) and the error itself for `error_QPE`. The reason being that we must have `error^2  >=  error_QPE^2 + (sum error_FPA)^2` and with our choices of errors esp. by setting `error_QPE = error`, this is being violated.